collections/src/main/scala/strawman/collection/immutable/LazyList.scala

```diff
@@ -220,6 +220,93 @@ sealed abstract class LazyList[+A] extends LinearSeq[A] with LazyListOps[A, Lazy
     if (this.isEmpty) z
     else tail.foldLeft(op(z, head))(op)
   }
+
+  override def toString: String = {
+    /** Write all defined elements of this iterable into given string builder.
+      *  The written text begins with the string `start` and is finished by the string
+      *  `end`. Inside, the string representations of defined elements (w.r.t.
+      *  the method `toString()`) are separated by the string `sep`. The method will
+      *  not force evaluation of undefined elements. A tail of such elements will be
+      * represented by a `"?"` instead.  A cyclic stream is represented by a `"..."`
+      * at the point where the cycle repeats.
+      *
+      * @param b The [[collection.mutable.StringBuilder]] factory to which we need
+      * to add the string elements.
+      * @param start The prefix of the resulting string (e.g. "LazyList(")
+      * @param sep The separator between elements of the resulting string (e.g. ",")
+      * @param end The end of the resulting string (e.g. ")")
+      * @return The original [[collection.mutable.StringBuilder]] containing the
+      * resulting string.
+      */
+    def toStringBuilder(b: StringBuilder, start: String, sep: String, end: String): StringBuilder = {
+      b append start
+      if (nonEmpty) {
+        if (headDefined) b append head else b append "?"
+        var cursor = this
+        if (cursor.tailDefined) {  // If tailDefined, also !isEmpty
+          var scout = cursor.tail
+          if (cursor ne scout) {
+            cursor = scout
+            if (scout.tailDefined) {
+              scout = scout.tail
+              // Use 2x 1x iterator trick for cycle detection; slow iterator can add strings
+              while ((cursor ne scout) && scout.tailDefined) {
+                if (cursor.headDefined) b append sep append cursor.head
+                cursor = cursor.tail
+                scout = scout.tail
+                if (scout.tailDefined) scout = scout.tail
+              }
+            }
+          }
+          if (!scout.tailDefined) {  // Not a cycle, scout hit an end
+            while (cursor.headDefined) {
```

Comment:
I think this condition is wrong. We want to progress even if the `head` hasn’t been evaluated, as long as the `tail` has been.